src/Umbraco.Infrastructure/Runtime/AppPluginsManifestWatcher.cs

```diff
@@ -20,45 +20,48 @@ public sealed class ManifestWatcher :
 
         // if configured and in debug mode, a ManifestWatcher watches App_Plugins folders for
         // package.manifest chances and restarts the application on any change
-        private Core.Manifest.ManifestWatcher _mw;
+        private ManifestWatcher _mw;
 
-        public ManifestWatcher(IHostingEnvironment hosting, ILoggerFactory loggerFactory, IHostingEnvironment hostingEnvironment, IUmbracoApplicationLifetime umbracoApplicationLifetime)
+        public AppPluginsManifestWatcher(IHostingEnvironment hosting, ILoggerFactory loggerFactory, IHostingEnvironment hostingEnvironment, IUmbracoApplicationLifetime umbracoApplicationLifetime)
         {
             _hosting = hosting;
             _loggerFactory = loggerFactory;
             _hostingEnvironment = hostingEnvironment;
             _umbracoApplicationLifetime = umbracoApplicationLifetime;
         }
 
-        public Task HandleAsync(UmbracoApplicationStarting notification, CancellationToken cancellationToken)
+        public void StartMonitoring()
         {
             if (_hosting.IsDebugMode == false)
             {
-                return Task.CompletedTask;
+                return;
             }
 
             var appPlugins = _hostingEnvironment.MapPathContentRoot(Constants.SystemDirectories.AppPlugins);
             if (Directory.Exists(appPlugins) == false)
             {
-                return Task.CompletedTask;
+                return;
             }
 
-            _mw = new Core.Manifest.ManifestWatcher(_loggerFactory.CreateLogger<Core.Manifest.ManifestWatcher>(), _umbracoApplicationLifetime);
+            _mw = new ManifestWatcher(_loggerFactory.CreateLogger<ManifestWatcher>(), _umbracoApplicationLifetime);
```

Comment:
Should there be more than one `ManifestWatcher` running at a time? If not' why not inject that directly?

Incidentally, that class needs cleaning up. The `DisposableObjectSlim` class it inherits is a mess since it explicitly states it is for objects with no unmanaged objects then goes on to call the `GC.SuppressFinalize(this)` method for a non-existing finalizer.

Reply:
🤷

Reply:
In terms of handling singleton object + transient event, better?

Reply:
Haha.... Looking at the code, I would say no. The `ManifestWatcher` should be injected as a singleton and your `INotificationHandler` simply adds the AppPlugins folder to that watcher via `Start` (_poorly named if the methods can be called more that once_). It appears to do things like check the directory/file exists already.

Reply:
`DisposableObjectSlim` is just following MS practices. GC.SupressFinalize doesn't do anything if there's no finalizer, see https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose?redirectedfrom=MSDN#the-dispose-method ... but sure that could be removed. In any case, this should still use the correct IDisposable pattern and not just implement Dispose without checking if it's already been disposed. 

Using VS when ctrl + . and choosing 'implement interface with IDispoable pattern' will also add this by default:

```cs
protected virtual void Dispose(bool disposing)
{
    if (!_disposedValue)
    {
        if (disposing)
        {
            // TODO: dispose managed state (managed objects)
        }

        // TODO: free unmanaged resources (unmanaged objects) and override finalizer
        // TODO: set large fields to null
        _disposedValue = true;
    }
}

// // TODO: override finalizer only if 'Dispose(bool disposing)' has code to free unmanaged resources
// ~ModelsBuilderComponent()
// {
//     // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
//     Dispose(disposing: false);
// }

public void Dispose()
{
    // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
    Dispose(disposing: true);
    GC.SuppressFinalize(this);
}
```

Reply:
`ManifestWatcher` should be sealed and contains no unmanaged types so the call (_which isn't free_) is redundant. It's included in the pattern to ensure that it's not overlooked in cases when it is required. 

IMO classes like `DisposableSlim` are dangerous. If you need to manage the lifetime of objects within your object you should do so within that object only.

Reply:
Incidentally here's probably the most in-depth article on the subject. Peer reviewed by several heavy-hitters at Microsoft.

http://joeduffyblog.com/2005/04/08/dg-update-dispose-finalization-and-resource-management/

Reply:
If kept, `DisposableSlim` should be updated to use the standard conventions/practices and naming. 

> so the call (which isn't free) is redundant

I agree, can be removed

Out of interest, here's the benchmark. It's not free but I don't think anyone is going to notice since its not critical code :)

``` ini

BenchmarkDotNet=v0.11.3, OS=Windows 10.0.18363
Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
  [Host]     : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 32bit LegacyJIT-v4.8.4250.0
  DefaultJob : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 32bit LegacyJIT-v4.8.4250.0


```
|                  Method |     Mean |     Error |    StdDev | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|------------------------ |---------:|----------:|----------:|------------:|------------:|------------:|--------------------:|
|    WithSuppressFinalize | 4.913 ns | 0.1345 ns | 0.1258 ns |      0.0038 |           - |           - |                12 B |
| WithoutSuppressFinalize | 2.304 ns | 0.0649 ns | 0.0575 ns |      0.0038 |           - |           - |                12 B |


> IMO classes like DisposableSlim are dangerous

Sure, I'm not wedded to this class and happy to have it removed and implement the pattern manually wherever necessary. Just want to mention the trade-off which is that anywhere IDisposable is needed, the pattern will typically be generated from VS/Ryder snippets which means there will be plenty of people leaving that call alone since VS's comments on that code block say `Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method`

`DisposableSlim` is only for simple non-finalizer requirements and should be fine even given all the dos/donts of that article since the complexity and challenges arise from when finalization is required.

If anyone feels strongly about the removal of the class, feel free to go for it

Reply:
By default resharper doesn't bother with Dispose(bool) unless the class has a finalizer or you click a checkbox for subclass/unmanaged resources.

Reply:
Same with VS, there's the option to `implement interface with IDisposable pattern` or just implement the interface. For now I was just fixing up DisposableObjectSlim for the interim but noticed this when we remove suppress finalize 

![image](https://user-images.githubusercontent.com/1742685/102943717-f0fd7a80-450c-11eb-9b53-b37e7c8a48f1.png)

So we either have to disable that warning in code or add config to suppress this rule.

Reply:
Or delete `DisposableObjectSlim`

Reply:
What I'm saying though is that anywhere IDisposable is implemented, our current analysis rules will bark at us for not using GC.SuppressFinalize.